Moved batch operations APIs to TableAsyncClient and TableClient.

vcolin7 opened the pull request.

joheredi reviewed (commented):
Do you have any samples that show how to use the batch from the TableClient?

Also IIRC one of the comments in the arch-board was to refer to the batch as a `transaction`. In the TS SDK I have renamed all instances of batch to only be `transaction`. I see in the changelog that there are still a few references to batch

vcolin7 commented:
> 
> 
> Do you have any samples that show how to use the batch from the TableClient?
> 
> Also IIRC one of the comments in the arch-board was to refer to the batch as a `transaction`. In the TS SDK I have renamed all instances of batch to only be `transaction`. I see in the changelog that there are still a few references to batch

@joheredi Do you mean this part?

> In their place, batch operations can now be submitted via the following methods:

If so, I think I could rephrase it a bit to something like:

> In their place, a batch of operations can now be submitted as a transaction via the following methods:

Else, could you be a bit more specific? I think we can still use the word batch as long as it's clear it's a batch of operations that will be treated in a transactional fashion. For example, .NET has documentation [like this](https://github.com/Azure/azure-sdk-for-net/blob/1f1e21a699cb82acc45f9b566a5e1c10a0f36720/sdk/tables/Azure.Data.Tables/src/TableClient.cs#L1277) on their `submitTransaction()` methods.

As for samples of how to use this, I will add them in a future PR along with samples for other Tables scenarios. In the meantime, you can look at [the Test files I modified in this PR](https://github.com/Azure/azure-sdk-for-java/blob/e703727a97de18befb8504a8728a142537baf5fe/sdk/tables/azure-data-tables/src/test/java/com/azure/data/tables/TablesAsyncClientTest.java#L719) for an example.

christothes commented on sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/models/TableTransactionAction.java:
```diff
@@ -0,0 +1,66 @@
+     * @param entity The table entity to which the {@code actionType} will be applied.
+     * @param ifUnchanged When {@code true}, the ETag of the provided entity must
+     * match the ETag of the entity in the Table service. If the values do not match, the update will not occur and
+     * an exception will be thrown. This value is only applied for update operations.
```
This should apply to Delete operations also.

vcolin7 replied:
You mean having the `ifUnchanged` argument?

christothes replied:
correct - I believe that operation can take an etag parameter.

vcolin7 replied:
Oh yes, we take that in our `TableClient` and `TableAsyncClient`. I will make sure the TableTransactionAction documentation reflects the actual behavior of when a user passes an `eTag` that does not match. By the way, I saw a bunch of other services name the ETag paramter `ifMatch`, should we consider doing the same?

vcolin7 replied:
Also, when looking at .NET's code I couldn't help but wonder what would happen if a user creates a `TableTransactionAction` with an ETag that's different than the one in the `Entity` field? We can encounter the same issue in Java if I were to add such a parameter to our constructor.

vcolin7 replied:
Talked to Chris separately and he clarified that the value specified in the input parameters is the one considered. The ETag in a `TableEntity` gets ignored.

christothes commented on sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/TableAsyncClient.java:
```diff
@@ -1076,4 +1079,168 @@ private AccessPolicy toAccessPolicy(TableAccessPolicy tableAccessPolicy) {
+     * methods in {@link TableAsyncClient} to understand the conditions that may cause a given operation to fail.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public synchronized Mono<List<BatchOperationResponse>> submitTransaction(List<TableTransactionAction> transactionalBatch) {
```
Should we call this parameter `transactionActions`?

vcolin7 replied:
We could, especially if we want to remove references to the work `batch`, but .NET would have to do so as well. Honestly I just grabbed the name .NET was using to be consistent :P

christothes replied:
.NET names it transactionActions

https://github.com/Azure/azure-sdk-for-net/blob/74293800392e75c160d841784e01ae74b1bd93a9/sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs#L45

christothes commented on sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/TableAsyncClient.java:
```diff
@@ -1076,4 +1079,168 @@ private AccessPolicy toAccessPolicy(TableAccessPolicy tableAccessPolicy) {
+            String message = "An operation within the batch failed, the transaction has been rolled back.";
+
+            if (failedOperation != null) {
+                message += " The failed operation was: " + failedOperation;
```
Should we make the failedOperation value a property on the exception?

vcolin7 replied:
I don't think so, we are already appending the contents of the failed operation in the exception message. It would look something like this:

```java
"CreateEntity{partitionKey='" + entity.getPartitionKey() + "', rowKey='" + entity.getRowKey() + "'}";
```

christothes replied:
Here is what we did in .NET after some discussion in the last arch board review:

https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/tables/Azure.Data.Tables/src/TableTransactionFailedException.cs#L20

vcolin7 replied:
Also, the failed operation type is the internal `BatchOperation`. We would need to either introduce a public type to represent these operations or just use their `String` representation. We would probably need to introduce a new type of exception, the one being used right now is very generic and used across the whole Tables SDK and I'm not a fan of adding a new property to it only for transactional operations.

vcolin7 replied:
Ended up introducing a new type of exception for this: `TableFailedTransactionException`.

christothes commented on sdk/tables/azure-data-tables/src/test/resources/session-records/TablesAsyncClientTest.submitTransactionAsync.json:
```diff
@@ -0,0 +1,77 @@
+{
+  "networkCallRecords" : [ {
```
Just curious why this affected recordings

vcolin7 replied:
It didn't, I just ran the tests against the service in record mode to ensure the contents looked the same as what we had in our recordings already. Since it looked good, I decided to grab the new recordings instead, considering I changed the name of the test methods and all that.

joheredi commented on sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/implementation/models/MultipartPart.java:
```diff
@@ -16,7 +16,7 @@
 
     public MultipartPart(String boundaryPrefix) {
-        this.boundary = boundaryPrefix + "_" + UUID.randomUUID().toString();
+        this.boundary = boundaryPrefix + "_" + UUID.randomUUID();
```
Oh cool, as I understand this, Java calls toString() automatically when concatenating strings. Is that correct?

vcolin7 replied:
That's correct.

joheredi commented on sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/models/TableTransactionActionType.java:
```diff
@@ -0,0 +1,44 @@
+     * Upsert the entity in {@link TableEntityUpdateMode#REPLACE} mode. This is equivalent to
+     * {@link TableClient#upsertEntity(TableEntity)}.
+     */
+    UPSERT_REPLACE,
```
In JS we have a single action for Upsert and another one for Update, and pass the update mode as an optional element. Is it possible to do something like that in Java?

```java
new TableTransactionAction(TableTransactionActionType.UPSERT, firstEntity, "Merge")
new TableTransactionAction(TableTransactionActionType.UPSERT, firstEntity, "Replace")
```

vcolin7 replied:
To be honest, in this case we are following what .NET did here, but after talking with our Java deputy architect we felt like having a constructor that takes a `TableEntity`, `TableTransactionAction`, `TableTransactionActionType` and an `eTag` would make it so that it could only be used for update/upsert actions, restricting its use. In general it just complicates things more.

joheredi reviewed (approved):
Changes look good! Left a couple of questions, mostly to learn how Java works in those situations.
